src/Bicep.Core/Workspaces/SourceFileFactory.cs

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using Azure.Deployments.Core.Configuration;
+using Azure.Deployments.Core.Constants;
+using Azure.Deployments.Core.Definitions.Schema;
+using Azure.Deployments.Templates.Engines;
+using Bicep.Core.FileSystem;
+using Bicep.Core.Parsing;
+using Bicep.Core.Syntax;
+using Bicep.Core.Text;
+using Newtonsoft.Json.Linq;
+
+namespace Bicep.Core.Workspaces
+{
+    public static class SourceFileFactory
+    {
+        public static ISourceFile CreateSourceFile(Uri fileUri, string fileContents, bool isEntryFile) => isEntryFile
+            ? CreateBicepSourceFile(fileUri, fileContents)
+            : CreateSourceFile(fileUri, fileContents);
+
+        public static ISourceFile CreateSourceFile(Uri fileUri, string fileContents) =>
+            PathHelper.HasExtension(fileUri, LanguageConstants.JsonFileExtension) ||
+            PathHelper.HasExtension(fileUri, LanguageConstants.JsoncFileExtension) ||
+            PathHelper.HasExtension(fileUri, LanguageConstants.ArmTemplateFileExtension)
+                ? CreateArmTemplateSourceFile(fileUri, fileContents)
+                : CreateBicepSourceFile(fileUri, fileContents);
+
+        public static SyntaxTree CreateBicepSourceFile(Uri fileUri, string fileContents)
+        {
+            var parser = new Parser(fileContents);
+            var lineStarts = TextCoordinateConverter.GetLineStarts(fileContents);
+            
+            return new SyntaxTree(fileUri, lineStarts, parser.Program());
+        }
+
+        public static ArmTemplateFile CreateArmTemplateSourceFile(Uri fileUri, string fileContents)
+        {
+            try
+            {
+                var template = TemplateEngine.ParseTemplate(fileContents);
+
+                ValidateTemplate(template);
+
+                var templateObject = JObject.Parse(fileContents, new JsonLoadSettings
```

Comment:
>JsonLoadSettings [](http://example.com/codeflow?start=69&length=16)

Does the default duplicate property behavior match the runtime? Should we set it explicitly?

Reply:
Yes it matches (the runtime does not specify `DuplicatePropertyNameHandling`).